### PR TITLE
Fix MSAN issue caused by incorrect date format.

### PR DESCRIPTION
**Pull request by JackyWoo**
see https://github.com/ClickHouse/ClickHouse/pull/67013#issuecomment-2275550343

https://s3.amazonaws.com/clickhouse-test-reports/67013/6e7587e9b3dbc791f6970fe23703c1d65aa4b7f0/stateless_tests__msan__%5B3_4%5D.html

### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix MSAN issue caused by incorrect date format. 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit ca4041847e4aa8acccd6ea31c0a18f2160c0dc7a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68105/ca4041847e4aa8acccd6ea31c0a18f2160c0dc7a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/21f8b03789220648948a0d545989d77d9183a927/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/IO/ReadHelpers.cpp [hunk 1]
```diff
@@ -1402,7 +1402,7 @@ ReturnType readDateTimeTextFallback(time_t & datetime, ReadBuffer & buf, const D
             s_pos[size] = 0;
 
             if constexpr (throw_exception)
-                throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Cannot parse DateTime {}", s);
+                throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Cannot parse DateTime {}", String(s, already_read_length));
```
Comment (antonio2368) [inline src/IO/ReadHelpers.cpp:1405]:
```suggestion
                throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Cannot parse DateTime {}", std::string_view(s, already_read_length + size));
```

also, you shoulbe be able to remove `s_pos[size] = 0` above

File: src/IO/ReadHelpers.cpp [hunk 2]
```diff
@@ -1432,7 +1432,7 @@ ReturnType readDateTimeTextFallback(time_t & datetime, ReadBuffer & buf, const D
                 s_pos[size] = 0;
 
                 if constexpr (throw_exception)
-                    throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Cannot parse time component of DateTime {}", s);
+                    throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Cannot parse time component of DateTime {}", String(s, size));
```
Comment (antonio2368) [inline src/IO/ReadHelpers.cpp:1435]:
```suggestion
                    throw Exception(ErrorCodes::CANNOT_PARSE_DATETIME, "Cannot parse time component of DateTime {}", std::string_view(s, size));
```

you should be able to remove `s_pos[size] = 0;` here also

**Review by alexey-milovidov — APPROVED** [review]

**Comment by JackyWoo** [comment]
@antonio2368 can you look at the PR? The fa

**Comment by yariks5s** [comment]
Hi @JackyWoo! 
1. The integration tests really seem to be not related to your PR. You can try to merge your branch with the master as we should have this test fixed already.
2. Bugfix validation checks if the test that you wrote was failing without your changes. As it is running without sanitizer there, it passes without your changes, and that causes the failure of bugfix validation. What you can do is write a small integration test with the same queries, but there you will check the full error message, so in this case validation will be passed.

**Review by yariks5s — APPROVED** [review]

**Comment by JackyWoo** [comment]
Thank you

**Comment by JackyWoo** [comment]
@yariks5s all checks hasve passed.

**Pull request merged**